### PR TITLE
Refactor: `BalancerMinter`

### PR DESCRIPTION
**Pull request by jubeira**
# Description

Refactor common functionality for `BalancerMinter` into a base contract, paving the way for a pseudo minter.
This PR doesn't introduce any new changes in functionality; it just re-arranges the code for later reuse.

Builds on top of #2306

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [x] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

See #2289

### CODE REVIEW

File: pkg/liquidity-mining/contracts/BalancerMinterBase.sol [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+     */
+    function mint(address gauge) external override nonReentrant returns (uint256) {
+        return _mintFor(gauge, msg.sender);
+    }
+
+    /**
+     * @notice Mint everything which belongs to `msg.sender` across multiple gauges
+     * @param gauges List of `LiquidityGauge` addresses
+     */
+    function mintMany(address[] calldata gauges) external override nonReentrant returns (uint256) {
```
Comment (EndymionJkb) [inline pkg/liquidity-mining/contracts/BalancerMinterBase.sol:68]: Could use /// @inheritdoc IBalancerMinter for all these

File: pkg/interfaces/contracts/liquidity-mining/ILMGetters.sol [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity >=0.7.0 <0.9.0;
+
+import "./IBalancerTokenAdmin.sol";
+import "./IGaugeController.sol";
+
+interface ILMGetters {
```
Comment (jubeira) [inline pkg/interfaces/contracts/liquidity-mining/ILMGetters.sol:20]: If you have a better name for this thin interface I'm all ears :)

**Review by nventuro — APPROVED** [review]

File: pkg/interfaces/contracts/liquidity-mining/ILMGetters.sol (hunk 2 above)
Comment (jubeira) [reply pkg/interfaces/contracts/liquidity-mining/ILMGetters.sol:20]: We use the getters e.g. in `StakelessGauge`. So if these are not in an interface, we'd need to send the implementation as arguments to be able to use them.

File: pkg/liquidity-mining/contracts/BalancerMinter.sol [hunk 3]
```diff
@@ -162,53 +118,29 @@ contract BalancerMinter is IBalancerMinter, ReentrancyGuard, EOASignaturesValida
...
+    function _mintFor(address gauge, address user) internal virtual returns (uint256 tokensToMint);
 
-        ILiquidityGauge(gauge).user_checkpoint(user);
-        uint256 totalMint = ILiquidityGauge(gauge).integrate_fraction(user);
-        tokensToMint = totalMint.sub(_minted[user][gauge]);
-
-        if (tokensToMint > 0) {
-            _minted[user][gauge] = totalMint;
-            emit Minted(user, gauge, totalMint);
-        }
```
[the remark this answers is not in the document]
Comment (jubeira) [reply pkg/liquidity-mining/contracts/BalancerMinter.sol:193]: hmm not sure if I understand the comment, but the `if` evaluates the tokens to mint, and the event logs the total. Do you think the `if` should be in `_setMinted`?

**Comment by jubeira** [comment]
Thanks for the review!
I'll hold on merging until we approve #2306

**Pull request merged**